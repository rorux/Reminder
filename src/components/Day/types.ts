import { TRecord } from '@store/records/types';

export type TDayProps = {
  number: null | number;
  records: Array<TRecord>;
};

export type TStylesDay = {
  backgroundColor?: string;
  cursor?: 'pointer';
};
