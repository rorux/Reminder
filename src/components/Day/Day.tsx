import React, { useContext } from 'react';
import { useDispatch } from 'react-redux';
import { styled } from '@mui/material/styles';
import Paper from '@mui/material/Paper';
import { grey, green } from '@mui/material/colors';
import { TDayProps, TStylesDay } from './types';
import { MyContext } from '@components/containers/Calendar/Calendar';
import { modalOpenAction } from '@store/app/actions';
import { HOLIDAYS_2022 } from '@utils/records/constants';

export const Item = styled(Paper)(({ theme }) => ({
  ...theme.typography.caption,
  padding: theme.spacing(1),
  textAlign: 'center',
  width: 32,
  height: 32,
}));

const Day: React.FC<TDayProps> = ({ number, records }) => {
  const contextValue = useContext(MyContext);
  const dispatch = useDispatch();

  const handleShowRecords = () => {
    if (number !== null) dispatch(modalOpenAction({ records, date: number }));
  };

  const styles: TStylesDay = {};

  if (records.length) {
    styles.backgroundColor = grey[300];
    styles.cursor = 'pointer';
  }

  if (contextValue.todayParse === number) {
    styles.backgroundColor = green[200];
  }

  if (number !== null && HOLIDAYS_2022.includes(number)) {
    styles.color = 'red';
  }

  if (contextValue.todayParse === number && records.length && number !== null) {
    dispatch(modalOpenAction({ records, date: number }));
  }

  return (
    <Item
      data-test="item"
      elevation={0}
      sx={styles}
      onClick={records.length ? handleShowRecords : undefined}
    >
      {number !== null ? new Date(number).getDate() : ''}
    </Item>
  );
};

export default Day;
