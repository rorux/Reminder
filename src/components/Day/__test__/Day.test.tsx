import React from 'react';
import * as redux from 'react-redux';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Enzyme, { shallow } from 'enzyme';
import Day, { Item } from '../Day';

Enzyme.configure({ adapter: new Adapter() });

const record = {
  id: '1',
  title: 'title of monthly remind',
  period: 'monthly',
  monthday: '8',
  holidays: 'afterHoliday',
  days: [1642366800000, 1644872400000],
};

const setUp = () => shallow(<Day records={[record]} number={123} />);
describe('Day component', () => {
  let component: Enzyme.ShallowWrapper;
  let spyOnUseContext: any;
  let spyOnUseDispatch;
  let mockDispatch: jest.Mock;

  const mockUseContext = () => {
    spyOnUseContext.mockImplementationOnce(() => ({ todayParse: 123 }));
  };

  beforeEach(() => {
    spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
    mockDispatch = jest.fn();
    spyOnUseDispatch.mockReturnValue(mockDispatch);

    spyOnUseContext = jest.spyOn(React, 'useContext');
  });

  it('Should render Day component', () => {
    component = setUp();
    expect(component).toMatchSnapshot();
  });

  it('Item click should dispatch modalOpenAction', () => {
    component = setUp();
    const item = component.find(`[data-test='item']`);
    // @ts-ignore
    item.props().onClick();
    expect(mockDispatch).toHaveBeenCalled();

    // @ts-ignore
    expect(item.props().sx.backgroundColor).toBe('#e0e0e0');
  });

  it('Item changes backgroundColor when Context has todayParse === number', () => {
    mockUseContext();
    component = setUp();
    const item = component.find(`[data-test='item']`);

    // @ts-ignore
    expect(item.props().sx.backgroundColor).toBe('#a5d6a7');
    // @ts-ignore
    expect(item.props().sx.cursor).toBe('pointer');
  });

  it('Item changes color when holiday', () => {
    mockUseContext();
    const component = shallow(
      <Day records={[record]} number={Date.parse(String(new Date(2022, 0, 1)))} />
    );
    const item = component.find(`[data-test='item']`);

    // @ts-ignore
    expect(item.props().sx.color).toBe('red');
  });

  it('useDispatch have not be called when records empty', () => {
    component = shallow(<Day records={[]} number={123} />);
    const item = component.find(`[data-test='item']`);
    expect(item.props().children).toBe(1);
    // @ts-ignore
    expect(item.props().onClick).toBeUndefined();
  });

  it('Date is empty when prop number is null', () => {
    component = shallow(<Day records={[]} number={null} />);
    const item = component.find(`[data-test='item']`);
    expect(item.props().children).toBe('');
  });
});

describe('Item component', () => {
  it('Should render Item component', () => {
    const component = shallow(<Item />);
    expect(component).toMatchSnapshot();
  });
});
