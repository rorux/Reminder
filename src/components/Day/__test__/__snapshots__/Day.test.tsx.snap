// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day component Should render Day component 1`] = `
<Styled(ForwardRef(Paper))
  data-test="item"
  elevation={0}
  onClick={[Function]}
  sx={
    Object {
      "backgroundColor": "#e0e0e0",
      "cursor": "pointer",
    }
  }
>
  1
</Styled(ForwardRef(Paper))>
`;

exports[`Item component Should render Item component 1`] = `
<ContextProvider
  value={
    Object {
      "insert": [Function],
      "inserted": Object {
        "n8odea": true,
      },
      "key": "css",
      "nonce": undefined,
      "registered": Object {
        "css-n8odea": "font-family:\\"Roboto\\", \\"Helvetica\\", \\"Arial\\", sans-serif;font-weight:400;font-size:0.75rem;line-height:1.66;letter-spacing:0.03333em;padding:8px;text-align:center;width:32px;height:32px;",
      },
      "sheet": StyleSheet {
        "_insertTag": [Function],
        "before": null,
        "container": undefined,
        "ctr": 0,
        "insertionPoint": undefined,
        "isSpeedy": false,
        "key": "css",
        "nonce": undefined,
        "prepend": undefined,
        "tags": Array [],
      },
    }
  }
>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-n8odea{font-family:\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif;font-weight:400;font-size:0.75rem;line-height:1.66;letter-spacing:0.03333em;padding:8px;text-align:center;width:32px;height:32px;}",
      }
    }
    data-emotion="css n8odea"
  />
  <ForwardRef(Paper)
    className="css-n8odea"
  />
</ContextProvider>
`;
